### PR TITLE
[InstCombine] Fold `select Cond, not X, X` into `Cond ^ X`

### PR DESCRIPTION
**Pull request by dtcxzyw**
See the following example:
```
define i1 @src(i64 %x, i1 %y) {
  %1526 = icmp ne i64 %x, 0
  %1527 = icmp eq i64 %x, 0
  %sel = select i1 %y, i1 %1526, i1 %1527
  ret i1 %sel
}

define i1 @tgt(i64 %x, i1 %y) {
  %1527 = icmp eq i64 %x, 0
  %sel = xor i1 %y, %1527
  ret i1 %sel
}
```
I find that this pattern is common in C/C++/Rust code base.
This patch folds `select Cond, Y, X` into `Cond ^ X` iff:
1. X has the same type as Cond
2. X is poison -> Y is poison
3. X == !Y

Alive2: https://alive2.llvm.org/ce/z/hSmkHS

It is an alternative to https://github.com/llvm/llvm-project/pull/90089.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 1]
```diff
@@ -3502,6 +3502,34 @@ static bool matchFMulByZeroIfResultEqZero(InstCombinerImpl &IC, Value *Cmp0,
...
+/// 2. X is true implies Y is false.
+/// 3. X is false implies Y is true.
+/// Otherwise, return false.
+static bool isKnownInversion(Value *X, Value *Y) {
+  // Handle X = icmp pred V, C1, Y = icmp pred V, C2.
+  Value *V;
+  Constant *C1, *C2;
+  ICmpInst::Predicate Pred1, Pred2;
+  if (!match(X, m_ICmp(Pred1, m_Value(V), m_Constant(C1))) ||
+      !match(Y, m_ICmp(Pred2, m_Specific(V), m_Constant(C2))))
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3516]: The two constants could be any value type. Wouldn't really req any change as you have the APInt check later on.

**Comment by goldsteinn** [comment]
Im somewhat partial to https://github.com/llvm/llvm-project/pull/90089. In general would prefer to use existing analysis logic (`isImpliedCondition...`) than this bespoke function.
Is there anything this function can do that other cant?

**Comment by dtcxzyw** [comment]
> Im somewhat partial to #90089. In general would prefer to use existing analysis logic (`isImpliedCondition...`) than this bespoke function. Is there anything this function can do that other cant?

No. I just use this simplified helper function to reduce compile-time overhead.
See https://github.com/llvm/llvm-project/pull/90089#discussion_r1616325307

**Comment by goldsteinn** [comment]
> > Im somewhat partial to #90089. In general would prefer to use existing analysis logic (`isImpliedCondition...`) than this bespoke function. Is there anything this function can do that other cant?
> 
> No. I just use this simplified helper function to reduce compile-time overhead. See [#90089 (comment)](https://github.com/llvm/llvm-project/pull/90089#discussion_r1616325307)

Okay, compile time is impact seems like a reasonable motivation.

**Comment by dtcxzyw** [comment]
Missed optimization from cvc5: https://alive2.llvm.org/ce/z/s3a6rd
```
define i1 @src (i1 %cmp23, i1 %pol) {
   %lnot = xor i1 %pol, true
   call void @use(i1 %lnot)
   %lnot.pol = select i1 %cmp23, i1 %lnot, i1 %pol
   ret i1 %lnot.pol
}

define i1 @tgt(i1 %cmp23, i1 %pol) {
   %lnot = xor i1 %pol, true
   call void @use(i1 %lnot)
   %lnot.pol = xor i1 %cmp23, %pol
   ret i1 %lnot.pol
}

declare void @use(i1)
```

**Comment by dtcxzyw** [comment]
> Missed optimization from cvc5: https://alive2.llvm.org/ce/z/s3a6rd
> 
> ```
> define i1 @src (i1 %cmp23, i1 %pol) {
>    %lnot = xor i1 %pol, true
>    call void @use(i1 %lnot)
>    %lnot.pol = select i1 %cmp23, i1 %lnot, i1 %pol
>    ret i1 %lnot.pol
> }
> 
> define i1 @tgt(i1 %cmp23, i1 %pol) {
>    %lnot = xor i1 %pol, true
>    call void @use(i1 %lnot)
>    %lnot.pol = xor i1 %cmp23, %pol
>    ret i1 %lnot.pol
> }
> 
> declare void @use(i1)
> ```

Without multi-use on xor: https://godbolt.org/z/TYxz1njab
```
INSTCOMBINE ITERATION #1 on src
ADD:   ret i1 %lnot.pol
ADD:   %lnot.pol = select i1 %cmp23, i1 %lnot, i1 %pol
ADD:   %lnot = xor i1 %pol, true
IC: Visiting:   %lnot = xor i1 %pol, true
IC: Visiting:   %lnot.pol = select i1 %cmp23, i1 %lnot, i1 %pol
ADD DEFERRED:   %1 = select i1 %cmp23, i1 true, i1 false
IC: Old =   %lnot.pol = select i1 %cmp23, i1 %1, i1 %pol
    New =   <badref> = xor i1 %pol, %lnot
ADD:   %lnot.pol = xor i1 %pol, %lnot
IC: ERASE   %2 = select i1 %cmp23, i1 %1, i1 %pol
ADD DEFERRED:   %1 = xor i1 %pol, true
IC: ERASE   %1 = xor i1 %pol, true
ADD:   %lnot = select i1 %cmp23, i1 true, i1 false
IC: Visiting:   %lnot = select i1 %cmp23, i1 true, i1 false
IC: Replacing   %lnot = select i1 %cmp23, i1 true, i1 false
    with i1 %cmp23
IC: Mod =   %lnot = select i1 %cmp23, i1 true, i1 false
    New =   %lnot = select i1 %cmp23, i1 true, i1 false
IC: ERASE   %lnot = select i1 %cmp23, i1 true, i1 false
IC: Visiting:   %lnot.pol = xor i1 %pol, %cmp23
IC: Visiting:   ret i1 %lnot.pol
```

As there is no way to remove one-use constraint in `InstCombinerImpl::foldSelectIntoOp`, should we handle this pattern in this PR?

**Comment by dtcxzyw** [comment]
Compile-time impact: http://llvm-compile-time-tracker.com/compare.php?from=4b5e0a1d12f0d624852c6d7c82207647ed5282aa&to=c7fb076906dff0f140462c093571e7f555d1271e&stat=instructions%3Au

**Comment by dtcxzyw** [comment]
Ping.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3516]: This is outstanding. Otherwise basically lgtm

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3516]: > The two constants could be any value type. Wouldn't really req any change as you have the APInt check later on.

`m_Specific(V)` implies the types of C1 and C2 are the same.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3516]: I think what @goldsteinn means is that you can replace the m_Constant with m_Value.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Done.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 2]
```diff
@@ -3996,5 +4023,11 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
     }
   }
 
+  // select Cond, !X, X -> xor Cond, X
+  // Note: We don't fold select Cond, Y, X -> X (iff X->Y & !X->!Y) here as
+  // it indicates that these two patterns should be canonicalized.
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:4028]: See https://github.com/llvm/llvm-project/pull/90089#discussion_r1588054750.

[the remark this answers is not in the document]
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:4028]: I don't think the comment is helpful with the new implementation, which does not perform implication checks.

**Comment by goldsteinn** [comment]
LGTM

**Pull request merged**